django/utils/encoding.py

```diff
@@ -196,20 +193,45 @@ def iri_to_uri(iri):
     return quote(force_bytes(iri), safe=b"/#%[]=:;$&()+,!?*@'~")
 
 
+# List of byte values that uri_to_iri is going to decode from percent encoding.
+# These are the unreserved characters from RFC 3986 and everything above 128,
+# because bytes >= 128 are part of multibyte unicode characters.
+_decode_ranges = [[45, 46, 95, 126], range(65, 91), range(97, 123), range(128, 256)]
+_hextobyte = {
+    (fmt % char).encode(): six.int2byte(char)
+    for range in _decode_ranges for char in range for fmt in ['%02x', '%02X']
+}
+
+
 def uri_to_iri(uri):
     """
     Converts a Uniform Resource Identifier(URI) into an Internationalized
     Resource Identifier(IRI).
 
-    This is the algorithm from section 3.2 of RFC 3987.
+    This is the algorithm from section 3.2 of RFC 3987, excluding step 4.
 
     Takes an URI in ASCII bytes (e.g. '/I%20%E2%99%A5%20Django/') and returns
     unicode containing the encoded result (e.g. '/I \xe2\x99\xa5 Django/').
     """
     if uri is None:
         return uri
     uri = force_bytes(uri)
-    iri = unquote_to_bytes(uri) if six.PY3 else unquote(uri)
+
```

Comment:
That does not seem very practical. This code is far enough removed from the original, that it is a lot easier to just look at it and see what it does than to understand how it differs from the original. The only concept that is the same is “split the string at "%" and then look up the two characters after the "%" in a map”. But even *why* this lookup is done is not the same anymore.